src/ddmd/dcache.d

```diff
@@ -0,0 +1,377 @@
+                    p = cast(Data*)(cast(void*)data + p.next);
+                    // Continue the pass to coalece the free blocks
+                    while (p.next != 0)
+                    {
+                        // remove old value if found
+                        if (0 == strncmp(p.key.ptr, key.ptr, key.length))
+                        {
+                            p.occupied = 0;
+                        }
+                        auto next = cast(Data*)(cast(void*)data + p.next);
+                        if (p.occupied == 0 && next.occupied == 0)
+                        {
+                            p.size = next.size + p.size;
+                            p.next = next.next;
+                        }
+                        else
+                            p = next;
+                    }
+                    return true;
+                }
+                offset = (offset + 1) & (toc.length-1);
+            }
+        }
+        
+    private:
+
+        void printToc()
+        {
+            printf("--- DCache TOC ---\n");
+            foreach(entry; toc)
+            {
+                if (entry.lastUsed != 0)
+                    printf("hash: %x index: %d lastUsed: %d\n", entry.hash32, entry.index, cast(int)entry.lastUsed);
+            }
+            printf("------------------\n");
+        }
+
+        void printBlocks()
+        {
+            Data* p = data;
+            printf("--- DCache Blocks ---\n");
+            for(;;)
+            {
+                printf("occupied = %d size = %d key = %s next = %d\n", p.occupied, p.size, p.key.ptr, p.next);
+                p = cast(Data*)(cast(void*)data + p.next);
+                if (p == data) break;
+            }
+            printf("---------------------\n");
+        }
+
+        static void[] mapRange(size_t size, int fd)
+        {
+            void * p = mmap(null, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+            if (p == MAP_FAILED)
+            {
+                perror("Failed to map");
+                abort();
+            }
+            return p[0..size];
+        }
+
+        static uint fnv1a(const(char)[] key)
+        {
+            uint hash = 0x811C9DC5;
+            foreach(char c; key)
+            {
+                hash = hash ^ c;
+                hash = hash * 0x01000193;
+            }
+            return hash;
+        }
+
+        void lock()
+        {
+            timespec timeout = timespec(0, 1_000_000); // 1M of nanoseconds - a millisecond
+            for (;;)
+            {
+                shared pid_t old = atomicLoad(*owner);
+                if (old != 0)
+                {
+                    // Use null signal to check presense of process
+                    if (kill(old, 0) >= 0)
+                    {
+                        // somebody is alive and has pid written to lock - wait
+                        nanosleep(&timeout, &timeout);
```

Comment:
Maybe use a shared lock on the file instead of busy polling?
I recently switched dub to use `flock`, this will also take care of dying processes, and as a bonus you could only lock certain regions.
https://github.com/dlang/dub/blob/255209da6491f98df602b121a8effe2c1fc94e69/source/dub/internal/utils.d#L62
https://github.com/dlang/phobos/blob/c1f616b618c459936c4b42753c7acf800f20d470/std/stdio.d#L1332

Reply:
Hm might work. Still the cache is a simple hash map and if there is heavy contention we've failed.
At the moment it caches may half a dozen of CTFE ecvaluations with threshold > 100ms to compute.

Reply:
> So this comment is wrong or what?
> You only need to do O_CREAT|O_RDWR here.

Okay